ky_expressions/expressions.h

```diff
@@ -0,0 +1,74 @@
+#ifndef KY_EXPRESSIONS_H
+#define KY_EXPRESSIONS_H
+
+#include <memory>
+
+namespace ky_expr {
+
+class ConstantExpression;
+class BinaryOperatorExpression;
+class ExpressionVisitor {
+public:
+  virtual void Visit(ConstantExpression &) = 0;
+  virtual void Visit(BinaryOperatorExpression &) = 0;
+};
```

Comment:
Thinking more carefully about this, I think this can be a concept and it does not require the virtual functions.
It could be handled by a concept, which I might look into :)

Reply:
It will remove the runtime cost at the expense of having to template-instantiate the expression hierarchy for each new visitor. It might be worth it depending on context -- compile time vs. runtime.